state.go

```diff
@@ -584,19 +585,75 @@ func NewUpdateErrorState(err menderError, update UpdateResponse) State {
 }
 
 func (ue *UpdateErrorState) Handle(ctx *StateContext, c Controller) (State, bool) {
-	// TODO error handling, repeat status update for as long as needed
-	c.ReportUpdateStatus(ue.update, statusFailure)
-
-	// TODO fetch logs from logger
-	c.UploadLog(ue.update, []LogEntry{
-		LogEntry{
-			Timestamp: time.Now().Format(time.RFC3339),
-			Level:     "error",
-			Message:   "update failed",
-		},
-	})
+	return NewUpdateStatusReportState(ue.update, statusFailure), false
+}
+
+// Wrapper for mandatory update state reporting, the state handler will attempt
+// to report state for a number of times. In case of recurring failure, the
+// update is deemed as failed.
+type UpdateStatusReportState struct {
+	CancellableState
+	update UpdateResponse
+	status string
+}
+
+func NewUpdateStatusReportState(update UpdateResponse, status string) State {
+	return &UpdateStatusReportState{
+		CancellableState: NewCancellableState(BaseState{
+			id: MenderStateUpdateStatusReport,
+		}),
+		update: update,
+		status: status,
+	}
+}
+
+func (usr *UpdateStatusReportState) Handle(ctx *StateContext, c Controller) (State, bool) {
+	if err := StoreStateData(ctx.store, StateData{
+		Id:           usr.Id(),
+		UpdateInfo:   usr.update,
+		UpdateStatus: usr.status,
+	}); err != nil {
+		log.Errorf("failed to store state data in update status report state: %v",
+			err)
+		// TODO: what now?
+	}
+
+	try := 0
+
+	for {
+		try += 1
+
+		log.Infof("attempting to report status %v of deployment to the backend, try %v",
+			usr.status, usr.update.ID, try)
+		if merr := c.ReportUpdateStatus(usr.update, usr.status); merr != nil {
+			log.Errorf("failed to report status %v: %v", usr.status, merr.Cause())
+			// TODO: until backend has implemented status reporting, this error cannot
+			// result in update being aborted. Once required API endpoint is available
+			// uncomment the code below
+
+			// wait for some time before trying again
+			// if wc := usr.Wait(c.GetUpdatePollInterval()); wc == false {
+			// 	return usr, true
+			// }
+		} else {
+			if usr.status == statusFailure {
+				log.Debugf("update failed, attempt log upload")
+				// TODO fetch logs from logger
+				c.UploadLog(usr.update, []LogEntry{
+					LogEntry{
+						Timestamp: time.Now().Format(time.RFC3339),
+						Level:     "error",
+						Message:   "update failed",
+					},
+				})
+			}
+		}
+
+		log.Debug("reporting complete")
+		break
```

Comment:
Won't this `break` always be hit? It looks like you're trying to do some repeated attempts to upload the log, but I cannot see that the loop can ever be executed more than once, not even with the currently commented code.